src/lib/SaferNames/Name.hs

```diff
@@ -316,168 +234,339 @@ data ListE (e::E) (n::S) = ListE { fromListE :: [e n] }
-  env' = injectEnvNamesL ext env
-  in RecEnv $ env' <>> frag
+withDistinct :: ScopeReader m => (Distinct n => m n a) -> m n a
+withDistinct cont = askScope >>= \case Scope _ -> cont
 
-injectEnvNamesL :: HasNamesE e => FreshExt n l -> Env i (e n) -> Env i (e l)
-injectEnvNamesL ext env = fromEnvE $ injectNamesL ext $ EnvE env
+-- === subst monad ===
+
+class ScopeReader2 m => SubstReader (v::E->E) (m::MonadKind2) | m -> v where
+  askSubst :: m i o (Env v i o)
+  withSubst :: Env v i' o -> m i' o a -> m i o a
+
+-- we could have `m::MP` but it's more work and we don't need it
+newtype SubstReaderT (m:: * -> *) (v::E -> E) (i::S) (o::S) (a:: *) =
+  SubstReaderT { runSubstReaderT' :: ReaderT (Scope o, Env v i o) m a }
+  deriving (Functor, Applicative, Monad)
+
+type SubstReaderM = SubstReaderT Identity
+
+runSubstReaderT :: Scope o -> Env v i o -> SubstReaderT m v i o a -> m a
+runSubstReaderT bindings env m = runReaderT (runSubstReaderT' m) (bindings, env)
+
+runSubstReaderM :: Scope o -> Env v i o -> SubstReaderM v i o a -> a
+runSubstReaderM scope env m = runIdentity $ runSubstReaderT scope env m
+
+instance MonadError e m => MonadError e (SubstReaderT m v i o) where
+  throwError e = SubstReaderT $ throwError e
+  catchError (SubstReaderT m) f = SubstReaderT $ catchError m $ runSubstReaderT' . f
+
+instance MonadFail m => MonadFail (SubstReaderT m v i o) where
+  fail s = SubstReaderT $ fail s
+
+instance Monad m => ScopeReader (SubstReaderT m v i) where
+  askScope = SubstReaderT $ asks fst
+  extendScope scopeFrag@(ScopeFrag m) (SubstReaderT (ReaderT f)) =
+    SubstReaderT $ ReaderT \(scope, subst) ->
+    f (appendScope scope scopeFrag, injectNames (nameMapNames m) subst)
+
+instance Monad m => SubstReader v (SubstReaderT m v) where
+  askSubst = SubstReaderT $ asks snd
+  withSubst subst (SubstReaderT (ReaderT f)) =
+    SubstReaderT $ ReaderT \(scope, _) -> f (scope, subst)
+
+extendSubst :: SubstReader v m
+            => EnvFrag v i i' o -> m i' o a -> m i o a
+extendSubst frag cont = do
+  env <- askSubst
+  withSubst (env <>> frag) cont
+
+dropSubst :: SubstReader (SubstVal sMatch atom) m => Typeable sMatch => Typeable atom
+          => m o o a
+          -> m i o a
+dropSubst cont = withSubst idSubst cont
+
+withFreshM :: ScopeReader m => Typeable s => InjectableE s
+           => s o
+           -> (forall o'. NameBinder s o o' -> m o' a)
+           -> m o a
+withFreshM ann cont = do
+  scope <- askScope
+  case scope of
+    Scope m ->
+      withFresh (nameMapNames m) \b' -> do
+        let ann' = inject b' ann
+        extendScope (ScopeFrag (singletonNameMap b' (IdE ann'))) $
+          cont b'
+
+lookupSubstM :: SubstReader v m => Name s i -> m i o (v s o)
+lookupSubstM name = (!name) <$> askSubst
+
+-- === alpha-renaming-invariant equality checking ===
+
+type AlphaEq e = AlphaEqE e  :: Constraint
+
+checkAlphaEq :: AlphaEqE e => MonadErr1 m => ScopeReader m
+             => e n -> e n -> m n ()
+checkAlphaEq e1 e2 = do
```

Comment:
We never needed to check alpha-equality before this change. Why now?

Reply:
We used to check alpha-equality as part of ordinary equality, via the `Eq` instance for `Abs`. You could do it that way with the new system too, though it's a bit more awkward because you have to conjure up a scope by querying free variables. But I decided against it for performance reasons. In the both the old system and the new one, doing alpha equality by implementing `Eq` means traversing the body many times when you have nested abstractions.